backend/src/payment/providers/checkout-v3.ts

```diff
@@ -0,0 +1,209 @@
+
+  var nonce = uuidv4();
+  var body = orderToCreateRequestBody(order, args);
+  var headers: CheckoutHeaders = {
+    'checkout-account': config.user,
+    'checkout-algorithm': 'sha256',
+    'checkout-method': 'POST',
+    'checkout-nonce': nonce,
+    'checkout-timestamp': moment().format()
+  }
+
+  var signature = sign(headers, body);
+
+  var httpHeaders = {
+    'content-type': 'application/json',
+    'charset': 'utf-8',
+    'signature': signature
+  }
+
+  let resp, err;
+
+  [resp, err] = await to(axios({
+      baseURL: 'https://api.checkout.fi',
+      method: 'post',
+      url: '/payments',
+      headers: httpHeaders,
+      data: body,
+      params: headers
+    }));
+
+  if (err) {
+    if (!verify(err.response.headers.signature, err.response.headers, err.response.data)) {
+          log.error('Response signature validation failed', {order_id: order.order_id});
+    }
+    log.error('Posting new payment failed', {error: {data: err.response.data, status: err.response.status, header: err.response.headers}, provider: PROVIDER});
+    return reject(err.response.data);
+  }
+
+  if (!verify(resp.headers.signature, resp.headers, resp.data)) {
+    return reject({name: 'Verification error', message: 'Signature verification failed'});
+  }
+
+  const createResp: payment.ICreateResponse = {
+    payment_id: resp.data.transactionId,
+    redirect_url: resp.data.href,
+    payment_url: resp.data.href,
+    payment_provider: PROVIDER,
+    payload: resp.data.providers
+  }
+  
+  return resolve(createResp);
+}
+
+// TODO: implement callback handlers
+export function handleSuccessCallback(req: express.Request): Promise<payment.ISuccessResponse> {
+  const headers = req.query;
+
+  if (! verify(headers.signature, headers)) {
+    return reject({name: 'Verification error', message: 'Signature verification failed'});
+  }
+
+  return resolve({
+    payment_id: headers['checkout-transaction-id'],
+    payment_provider: PROVIDER,
+  });
+}
+
+export function handleCancelCallback(req: express.Request): Promise<payment.ICancelResponse> {
+  const headers = req.query;
+
+  if (! verify(headers.signature, headers)) {
+    return reject({name: 'Verification error', message: 'Signature verification failed'});
+  }
+
+  return resolve({
+    payment_id: headers['checkout-transaction-id'],
+    payment_provider: PROVIDER,
+  });
+}
+
+export function checkStatus(payment_id: string, payment_url: string): Promise<payment.IStatusResponse> {
+
+
+  return resolve({
+    payment_id,
+    payment_url,
+    status: 'not-implemented',
+  });
+}
+
+function sign(headers: {[key: string]: any}, body?: CreateRequestBody): string {
+  var payloadArray =
+    Object.keys(headers)
+      .filter((value: string) => {
+        return /^checkout-/.test(value);
+      })
+      .sort()
+      .map((key) => [ key, headers[key] ].join(':'))
+
+  var payload: string;
```

Comment:
I'd inline this as something like
`var payload = payloadArray.concat(body ? JSON.stringify(body) : '').join("\n") `

Reply:
Done with inlining